server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateDataStreamService.java

```diff
@@ -329,7 +334,7 @@ static ClusterState createDataStream(
             template.getDataStreamTemplate().isAllowCustomRouting(),
             indexMode,
             lifecycle == null && isDslOnlyMode ? DataStreamLifecycle.DEFAULT : lifecycle,
-            template.getDataStreamTemplate().hasFailureStore() ? DataStreamOptions.FAILURE_STORE_ENABLED : DataStreamOptions.EMPTY,
+            dataStreamOptions,
```

Comment:
There's a behaviour in `main` right now which surprised me. If you have a template which explicitly sets `"failure_store": false`, the metadata for the created data stream has `DataStreamOptions.EMPTY` (no failure store setting), whereas I would expect it to be `DataStreamOptions.DISABLED` (explicitly disabled, like it was in the template).

I'm wondering whether your PR is going to change that. For the record, I think it would be good if it did! But it'd be useful for me to know either way, since it has implications for my work with the cluster setting.

Reply:
I understand it's surprising but it's on purpose. 

The `failure_store` field in the `DataStreamTemplate` is not optional which means that every data stream that has already been created has it set to false. If we implement it the way you are suggesting, it is like we are explicitly disabling the failure store for all existing data streams.

This will create a huge conflict with the cluster setting you are working on.

Finally, since the failure store is behind a feature flag, it is much more probable that the user intended it to be empty. It might be an inconvenience for the users that had early access and were testing it, but I do not think this qualifies as a blocker. 

Does this make sense?

Reply:
Hmm, trying to think this through... I guess my question is *why* is the `failure_store` field in the template not optional?

Suppose you have a template with `"failure_store": false`. When you create a data stream from that, it won't have a `failure_store` object in its options. Therefore, you'll still get the failure store enabled if the data stream name matches the cluster setting. To be consistent, we need to ensure that if we detect a failure before the data stream has been created, we also enable the failure store if the name matches the setting, even if the template has `"failure_store": false`. The *only* way, AFAICS, that you can disable the failure store is to (manually or automatically) create the data stream and then do a `PUT _data_stream/{name}/_options`.

That may be okay, though I still worry it may surprise users. It also doesn't align with the PRD, which currently says this:

> Users can disable the failure store for individual data streams or via (component) templates if it’s enabled via the global setting.

We won't be supporting the 'via (component) templates' bit. So if we're doing this, I think we should check with Fabio again.

AFAICS, the only way to implement the PRD as written would be if `failure_store` was optional in the template, so that we can override the cluster setting by specifying either true or false, and default to what the cluster setting says by omitting it.

Reply:
Happy to move this conversation to Slack, or even Zoom, if that's easier.

Reply:
I think there is a misunderstanding here. There are certain things that make this piece of code ok. 

- The feature is behind a feature flag, so no production user should have used this. There are no bwc concerns.
- The feature is not usable, but it does have some side-effects, such as it introduced a `boolean` flag in the `DataStreamTemplate` that defaults to false and it's serialised.
- Finally, configuring the failure store in the `DataStreamTemplate` does not satisfy anymore the requirements because it does not support template composition.

So, this PR is meant to "permanently" move the configuration from the `DataStreamTemplate` to the `Template` as part of the `DataStreamOptions`. Furthermore, now that we know that we want a 3 state flag to enable the failure store, this is also taken into account.

So let's answer your questions:

> Hmm, trying to think this through... I guess my question is why is the failure_store field in the template not optional?

As far as I understand there was no requirement for a default state when it was first designed 1+ ago. 

> Suppose you have a template with "failure_store": false. When you create a data stream from that, it won't have a failure_store object in its options. 

`"failure_store": false` is not possible configuration anymore, so that's not a concern, right?

> AFAICS, the only way to implement the PRD as written would be if failure_store was optional in the template, so that we can override the cluster setting by specifying either true or false, and default to what the cluster setting says by omitting it.

+1 and that's how we are implementing it here, or it least that's what I intended, if you think I missed something let me know. Also, this why you are blocked by this PR to finish your work.

Reply:
I think I've misunderstood parts of your previous comment and whether you were describing the current state or the state after this PR 😓... We can jump on a call to discuss, I'm just going to jot down some notes.

In `main` right now, if I do this (just showing the relevant bit):

```
PUT _index_template/my-index-template
{
  "index_patterns": ["my-data-stream*"],
  "data_stream": {
    "failure_store": false
  },
  "composed_of": [ "my-mappings", "my-settings" ],
  "priority": 500
}

POST my-data-stream/_doc
{
  "@timestamp": "2099-05-06T16:21:15.000Z",
  "message": "192.0.2.42 - - [06/May/2099:16:21:15 +0000] \"GET /images/bg.jpg \
HTTP/1.0\" 200 24736"
}

GET _data_stream/my-data-stream/_options
```

the last call returns this:
```
{
  "data_streams" : [
    {
      "name" : "my-data-stream"
    }
  ]
}
```

IOW, setting `"failure_store": false` in the index template results in the data stream options being empty (not having `failure_store.enabled` set to `false`).

The behaviour is the same if I do

```
PUT _index_template/my-index-template
{
  "index_patterns": ["my-data-stream*"],
  "data_stream": {
  },
  "composed_of": [ "my-mappings", "my-settings" ],
  "priority": 500
}
```

instead, i.e. I omit the `failure_store` property from the template. Also, if I then `GET` the template definition, it does have `"failure_store" : false`, i.e. it got defaulted to that when I `PUT` the template.

The implication for my work is that, because we don't distinguish between `"failure_store": false` in the index template and that property not being present, we can't make the former override the cluster setting (which the PRD currently says we should).

Hopefully we're on the same page so far!

Now, I *thought* what you were saying (in your comment starting "I understand it's surprising but it's on purpose") is that it would behave the same way after your PR, just with the stuff moved from the index template to the component templates. If that was the case, we'd have the same problem.

But from your later comments, it sounds like you're actually saying is that we *will* distinguish between setting `failure_store` to false in the template and not setting it, and that the former will result in the data stream being created with `failure_store.enabled` explicitly set to false in its options. Which is what I was originally asking about! If that's the case, then I think we're all good, and sorry for the confusion.

Reply:
> But from your later comments, it sounds like you're actually saying is that we will distinguish between setting failure_store to false in the template and not setting it, and that the former will result in the data stream being created with failure_store.enabled explicitly set to false in its options. 

That is correct.